### PR TITLE
[WIP] Add proper versioning of MonoAndroid framework.

### PR DESCRIPTION
**Pull request by dmunch**
So far no versioning for the MonoAndroid framework has been handled. However, I've been running into the case of different version constraints on transitive dependencies depending on the MonoAndroid framework version. As an example (and actually the concrete case I'm running into) take the nuspec of Xamarin.Forms.

```
      <group targetFramework="MonoAndroid1.0">
        <dependency id="Xamarin.Android.Support.v4" version="[23.3.0]" />
        <dependency id="Xamarin.Android.Support.Design" version="[23.3.0]" />
        <dependency id="Xamarin.Android.Support.v7.AppCompat" version="[23.3.0]" />
        <dependency id="Xamarin.Android.Support.v7.CardView" version="[23.3.0]" />
        <dependency id="Xamarin.Android.Support.v7.MediaRouter" version="[23.3.0]" />
      </group>
      <group targetFramework="MonoAndroid7.0">
        <dependency id="Xamarin.Android.Support.v4" version="23.3.0" />
        <dependency id="Xamarin.Android.Support.Design" version="23.3.0" />
        <dependency id="Xamarin.Android.Support.v7.AppCompat" version="23.3.0" />
        <dependency id="Xamarin.Android.Support.v7.CardView" version="23.3.0" />
        <dependency id="Xamarin.Android.Support.v7.MediaRouter" version="23.3.0" />
      </group>
```

In my case I'd like to update the Xamarin.Android.* packages to a higher version. As a first step, and what worked pretty good and straight forward so far, I implemented proper versioning of the MonoAndroid framework. This is what this PR currently contains.

However, even with the proper MonoAndroid framework version handling paket still is not able to resolve the following `paket.dependencies` file. 

```
framework: monoandroid70

source https://nuget.org/api/v2

nuget Xamarin.Forms == 2.3.4.247

nuget Xamarin.Android.Support.v4 == 25.3.1
nuget Xamarin.Android.Support.v7.RecyclerView == 25.3.1
nuget Xamarin.Android.Support.Design == 25.3.1
nuget Xamarin.Android.Support.v7.CardView == 25.3.1
```

From what I understood by reading the latest pull requests is that this involves framework restriction handling and has been refactored quite a lot in the last months. 

I'm currently wondering if this kind of behavior is actually possible, and if so, at which part of the code I'd have to look to be able to understand its mechanism and maybe come up with a possible solution. Any hint and help is much appreciated! 

Thanks a lot in advance, also for all the past work on paket which really is a joy to use!

### CODE REVIEW

**Comment by matthid** [comment]
Let me take a look :)

**Comment by dmunch** [comment]
Wait a second, I might have forgotten to bump `NuGetPackageCache.CurrentCacheVersion` at some point. I just retried with the rebased version on 5.0.0-rc011 and the mentioned example is resolved fine now.

**Comment by matthid** [comment]
sounds good. btw there is a switch in the test suite to update all tests. so you don't have to update manually. can't link right now as i'm on mobile... just keep an eye on the diff if the changes are sane

**Comment by matthid** [comment]
https://github.com/fsprojects/Paket/blob/master/integrationtests/Paket.IntegrationTests/InstallSpecs.fs#L27

Let me know if you need help.

**Comment by dmunch** [comment]
Thanks for the hint! I wont be able to touch this during the week-end, but I should have updated test-cases and final modifications ready during next week. I've aleady been using paket with those modifications in our solution with the desired results, so I think I'm on the right way here.

File: src/Paket.Core/Versioning/FrameworkHandling.fs [hunk 1]
```diff
@@ -276,6 +276,46 @@ type SilverlightVersion =
...
+    | V5
+    | V6
+    | V7
+    | V71
+    member this.ShortString() =
+        match this with
+        | MonoAndroidVersion.V1    -> ""
+        | MonoAndroidVersion.V22   -> "2.2"
+        | MonoAndroidVersion.V23   -> "2.3"
+        | MonoAndroidVersion.V403  -> "4.03"
```
Comment (matthid) [inline src/Paket.Core/Versioning/FrameworkHandling.fs]: really? not 4.0.3?

File: src/Paket.Core/Versioning/FrameworkHandling.fs [hunk 2]
```diff
@@ -362,7 +402,17 @@ type FrameworkIdentifier =
...
+        | MonoAndroid MonoAndroidVersion.V22 -> [ MonoAndroid MonoAndroidVersion.V1 ]
+        | MonoAndroid MonoAndroidVersion.V23 -> [ MonoAndroid MonoAndroidVersion.V22 ]
+        | MonoAndroid MonoAndroidVersion.V403 -> [ MonoAndroid MonoAndroidVersion.V23 ]
+        | MonoAndroid MonoAndroidVersion.V41 -> [ MonoAndroid MonoAndroidVersion.V403 ]
+        | MonoAndroid MonoAndroidVersion.V43 -> [ MonoAndroid MonoAndroidVersion.V41 ]
+        | MonoAndroid MonoAndroidVersion.V44 -> [ MonoAndroid MonoAndroidVersion.V43 ]
+        | MonoAndroid MonoAndroidVersion.V5 -> [ MonoAndroid MonoAndroidVersion.V44 ]
+        | MonoAndroid MonoAndroidVersion.V6 -> [ MonoAndroid MonoAndroidVersion.V5 ]
+        | MonoAndroid MonoAndroidVersion.V7-> [ MonoAndroid MonoAndroidVersion.V6; DotNetStandard DotNetStandardVersion.V1_6 ]
+        | MonoAndroid MonoAndroidVersion.V71-> [ MonoAndroid MonoAndroidVersion.V7 ]
```
Comment (matthid) [inline src/Paket.Core/Versioning/FrameworkHandling.fs]: some whitespace (not important)

File: integrationtests/scenarios/i001333-dup-refs/before/ConsoleApplication1/ConsoleApplication1.fsprojtemplate [hunk 3]
```diff
@@ -89,7 +89,7 @@
...
...ifier) == '.NETStandard' And ($(TargetFrameworkVersion) == 'v1.1' Or $(TargetFrameworkVersion) == 'v1.2' Or $(TargetFrameworkVersion) == 'v1.3' Or $(TargetFrameworkVersion) == 'v1.4' Or $(TargetFrameworkVersion) == 'v1.5' Or $(TargetFrameworkVersion) == 'v1.6' Or $(TargetFrameworkVersion) == 'v2.0')) Or ($(TargetFrameworkIdentifier) == '.NETCoreApp' And ($(TargetFrameworkVersion) == 'v1.0' Or $(TargetFrameworkVersion) == 'v1.1' Or $(TargetFrameworkVersion) == 'v2.0')) Or ($(TargetFrameworkIdentifier) == 'MonoAndroid' And ($(TargetFrameworkVersion) == 'v1.0' Or $(TargetFrameworkVersion) == 'v2.2' Or $(TargetFrameworkVersion) == 'v2.3' Or $(TargetFrameworkVersion) == 'v4.03' Or $(TargetFrameworkVersion) == 'v4.1' Or $(TargetFrameworkVersion) == 'v4.3' Or $(TargetFrameworkVersion) == 'v4.4' Or $(TargetFrameworkVersion) == 'v5.0' Or $(TargetFrameworkVersion) == 'v6.0' Or $(TargetFrameworkVersion) == 'v7.0' Or $(TargetFrameworkVersion) == 'v7.1')) Or ($(TargetFrameworkIdentifier) == 'MonoTouch') Or ($(TargetFrameworkIdentifier) == 'Xamarin.iOS') Or ($(TargetFrameworkIdentifier) == 'Xamarin.Mac') Or ($(TargetFrameworkProfile) == 'Profile7') Or ($(TargetFrameworkProfile) == 'Profile44')">
```
Comment (matthid) [inline integrationtests/scenarios/i001333-dup-refs/before/ConsoleApplication1/ConsoleApplication1.fsprojtemplate]: again the 4.03 makes ma bit nervous here

**Comment by matthid** [comment]
Travis looks like another failure i've not seen jet...

File: src/Paket.Core/Versioning/FrameworkHandling.fs (hunk 1 above)
Comment (dmunch) [inline src/Paket.Core/Versioning/FrameworkHandling.fs]: Good catch, considering the [offical list of Xamarin.Android releases](https://developer.xamarin.com/releases/android/mono_for_android_4/mono_for_android_4.0.3/)  it's certainly 4.0.3 and my Xamarin Studio also gives this in the list of available target frameworks. Will fix that.

image

Comment (dmunch) [inline src/Paket.Core/Versioning/FrameworkHandling.fs]: Compared to the XS screenshot there's also some Versions still missing. Not sure if we should add them, or just keep waiting until people complain or need them. The thing is I just took the versions paket already recognized which where based on real-world versions I found during my usage of paket on Xamarin projects. Any thoughts?

Comment (forki) [inline src/Paket.Core/Versioning/FrameworkHandling.fs]: please add them if you already see these

**Pull request merged**